xlive/Blam/Engine/interface/hud.cpp

```diff
@@ -0,0 +1,140 @@
+#define k_hud_upscale_size 1.0f
+
+// you'd think we could use 1200 since that's the games normal max resolution, but seems 1200 still hides some text :(
+#define max_crosshair_text_scale 1080.0f * 0.0010416667f 
+// >= 1.25 will make text disappear
+#define max_ui_scale 1.049f 
+
+
+// Used to grab the default crosshair size before we modify it
+typedef void(__cdecl* update_hud_elements_display_settings_t)(int new_hud_size, int new_safe_area);
+update_hud_elements_display_settings_t p_update_hud_elements_display_settings;
+
+void __cdecl update_hud_elements_display_settings_hook(const int new_hud_size, const int new_safe_area)
+{
+	p_update_hud_elements_display_settings(new_hud_size, new_safe_area);
+	original_hud_scale = *Memory::GetAddress<float*>(0x46402C);
+	original_crosshair_text_scale = *Memory::GetAddress<float*>(0x464028);
+	
+	if (H2Config_hiresfix)
+	{
+		high_res_fix();
+	}
+
+	set_hud_size(1.0f);
+	set_crosshair_size(H2Config_crosshair_scale);
+}
+
+void high_res_fix()
+{
+	if (original_crosshair_text_scale > max_crosshair_text_scale)
+	{
+		// textScale = resolution_height * 0.0010416667
+		// but if it's too large the game won't render it some reason (max seems to be around 1.4, 1.3 when sword icon is in the text?)
+		// lets just set it to the largest known good value for now, until the cause of large text sizes being hidden is figured out
+		set_crosshair_and_text_size(max_crosshair_text_scale);
+	}
+
+	// UI_Scale was updated just before we were called, so lets fix it real quick...
+	float* g_ui_scale = Memory::GetAddress<float*>(0xA3E424);
+	if (*g_ui_scale > max_ui_scale)
+	{
+		// uiScale = resolution_height * 0.00083333335f
+		// at higher resolutions text starts being cut off (pretty much any UI_Scale above 1 will result in text cut-off)
+		// the render_text_wrapper_hook below fixes that by scaling the width of the element by UI_Scale (which the game doesn't do for some reason...)
+		// however fonts will stop being rendered if the UI_Scale is too large (>= ~1.25)
+		// so this'll make sure UI_Scale doesn't go above 1.249, but will result in the UI being drawn smaller
+		*g_ui_scale = max_ui_scale;
+	}
+}
+
+typedef int(__cdecl* render_text_wrapper_t)(rect2d* a1, rect2d* a2, int a3, int a4, int a5, float a6, int a7, int a8);
+render_text_wrapper_t p_render_text_wrapper;
+
+int __cdecl render_text_wrapper_hook(rect2d* a1, rect2d* a2, int a3, int a4, int a5, float a6, int a7, int a8)
+{
+	float g_ui_scale = *Memory::GetAddress<float*>(0xA3E424);
+	if (a2 && a2 != a1) // if a2 == a1 then this is a chapter name which scaling seems to break, so skip those ones
+	{
+		short width = a2->right - a2->left;
+		a2->right = a2->left + (short)((float)width * g_ui_scale); // scale width by UI_Scale, stops text elements from being cut-off when UI_Scale > 1
+	}
+	return p_render_text_wrapper(a1, a2, a3, a4, a5, a6, a7, a8);
+}
+
+void set_hud_size(const float size)
```

Comment:
`const` parameters should mainly be used for references or pointers to variables that shouldn't be modified

Reply:
otherwise is just extra syntax sugar

Reply:
changed